---
title: "Gene_Pheno_processing"
author: "Meg Hutch"
date: "January 9, 2020"
output: html_document
---

Objective: Pre-process genebody and phenotypic data for the Multi-Cancer deep learning analysis

```{r Packages}
library(dplyr)
library(caret)
library(DESeq2)
```

```{r Import Data}
# gene count data
gene_count <- read.delim2("C:/Users/User/Box Sync/Projects/Multi_Cancer_DL/Raw_Data/genebody_count_all.txt", header = TRUE, sep = "")

# gene annotation
gene_ano <- read.delim2("C:/Users/User/Box Sync/Projects/Multi_Cancer_DL/Raw_Data/genebody_anno.txt", header = TRUE, sep = "")

# Jan 23: updated phenotypic data
pheno <- readRDS("C:/Users/User/Box Sync/Projects/Multi_Cancer_DL/Raw_Data/DataInfoFilter.rds")

# Fragment length - final ids in the analysis are in this file
fl <- read.csv("C:/Users/User/Box Sync/Projects/Multi_Cancer_DL/Raw_Data/DataInfoFilter.csv")
```

Map the gene annotations to the gene body data

**strand + or -**
```{r}
gene_ano <- gene_ano %>% select(Geneid)

gene_count <- cbind(gene_ano, gene_count)
```

Modify gene count data - transpose to facilitate merging with phenotypic data 

```{r}
# transpose: rows to columns
gene_count_t <- t(gene_count) 

# convert first row to header
gene_col_names <- gene_count_t[1,]

# transpose column names
colnames(gene_count_t) <- gene_col_names

# remove the geneids from the first row
gene_count_t <- gene_count_t[-1,]

# convert to dataframe
gene_count_t <- as.data.frame(gene_count_t)

# temporarily covert rownames to first column
gene_count_t <- tibble::rownames_to_column(gene_count_t, "seq_num")
```

Merge gene counts with the patient data
```{r}
pts_colData <- merge(pheno, gene_count_t, by = "seq_num")
```

Select columns of interest (remove gene counts)
```{r}
pts_colData <- pts_colData %>% select(seq_num, Cat, gender, age)

# In the new pheno dataset - some of the healthy patients were given the label CRN to identify non-colon cancers. These are also healthy patients and so I am going to change the label as such
pts_colData$Cat[pts_colData$Cat == "CRN"] <- "HEA"
```

Add Fragment Length (this is the finalized patient list, so if a patient isn't included they will be removed from our gene_count_t df)

We have 342 final patients
```{r}
fl <- fl %>% select(seq_num, frag_mean)

# Merge with pts_colData
pts_colData  <- merge(pts_colData, fl, by = "seq_num") 
```

Keep only cancer samples + healthy patients (HBV is included in this dataset)
```{r}
# diagnoses to keep
dx_keep <- c("CRC", "ESCA", "HCC", "HEA", "STAD", "BRCA", "GBM")
dx_keep <- as.data.frame(dx_keep)
colnames(dx_keep) <- "Cat"

# keep only diagnoses of interst
pts_colData <- pts_colData %>% filter(Cat %in% dx_keep$Cat)
```

Only keep the seq_nums in the gene_count data that match the final patient cohort in the pts_colData
```{r}
pts_ids <- pts_colData$seq_num

gene_count <- gene_count[, pts_ids]
```

Normalize gene counts using deseq2 and normalizing by disease type

References: 

* https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html

* http://bioconductor.org/packages/release/bioc/manuals/DESeq2/man/DESeq2.pdf

* http://www.sthda.com/english/wiki/rna-seq-differential-expression-work-flow-using-deseq2

* https://lashlock.github.io/compbio/R_presentation.html
```{r}
# Convert pts_ColData columns to factors
pts_colData$seq_num <- as.factor(pts_colData$seq_num)
pts_colData$Cat <- as.factor(pts_colData$Cat)
pts_colData$gender <- as.factor(pts_colData$gender)

# relevel the Cat so that healhy status is the reference - this may not be relevant if design is not infleuncing the normalization - perhaps for for Differential Gene Expression
pts_colData$Cat <- as.factor(pts_colData$Cat)
pts_colData$Cat <- relevel(pts_colData$Cat, ref="HEA")

## Create DESeq2Dataset object where ~ 1 indicates no design
dds <- DESeqDataSetFromMatrix(countData = as.matrix(gene_count), colData = pts_colData, design = ~ Cat)

# run deseq - I believe this is the whole RNA seq DEG pipeline - this performs normalization but also includes other steps 
#dds <- DESeq(dds)

# This performs the median of ratios method of normalization - it appears that the design function does not influence the estimateSizeFactors normalization process - whether or not design = ~ Cat or design = ~ 1, results are the same.
dds <- estimateSizeFactors(dds)

# View normalization factor
sizeFactors(dds)

# To retrieve the normalized counts:
normalized_counts <- counts(dds, normalized=TRUE)

# transpose data
normalized_counts <- t(normalized_counts)

# convert to dataframe
normalized_counts <- as.data.frame(normalized_counts)

# convert row names to first column
normalized_counts <- tibble::rownames_to_column(normalized_counts, "seq_num")

# Add gene names
gene_names <- colnames(gene_count_t)

colnames(normalized_counts) <- gene_names
```

Merge phenotypic data with the gene_count data - mc_data frame will be our final dataframe
```{r}
mc_data <- inner_join(pts_colData, normalized_counts, by = "seq_num")
```

Next we also want to added the cfDNA concentration from the original Cancer info excel file. This was split into separate csvs for each diagnosis and include cfDNA concentration. I will process this data and add to the mc_data df.

```{r}
# breast cancer
brca <- read.csv("C:/Users/User/Box Sync/Projects/Multi_Cancer_DL/Raw_Data/brca_info.csv")
  
# colon cancer
crc <- read.csv("C:/Users/User/Box Sync/Projects/Multi_Cancer_DL/Raw_Data/crc_info.csv")
  
# glioblastoma
gbm <- read.csv("C:/Users/User/Box Sync/Projects/Multi_Cancer_DL/Raw_Data/gbm_info.csv")
  
# esophageal cancer
esca <- read.csv("C:/Users/User/Box Sync/Projects/Multi_Cancer_DL/Raw_Data/esca_info.csv")

# hepatitis B virus infection
hbv <- read.csv("C:/Users/User/Box Sync/Projects/Multi_Cancer_DL/Raw_Data/hbv_info.csv")

# liver cancer
hcc <- read.csv("C:/Users/User/Box Sync/Projects/Multi_Cancer_DL/Raw_Data/hcc_info.csv")

# stomach cancer
stad <- read.csv("C:/Users/User/Box Sync/Projects/Multi_Cancer_DL/Raw_Data/stad_info.csv")

# healthy 
hea <- read.csv("C:/Users/User/Box Sync/Projects/Multi_Cancer_DL/Raw_Data/hea_info.csv")
```

Merge all diagnoses together and extract seq_num, diagnosis, and cfDNA concentration. There are a few patients in these csvs that had incomplete information, and thus, it looks like some seq_num ids are repeated and given to other patients. 
```{r}
brca <- brca %>% select(seq_num, diagnosis, dilute_library_concentration)

crc <- crc %>% select(seq_num, diagnosis, dilute_library_concentration)

gbm <- gbm %>% select(seq_num, diagnosis, dilute_library_concentration)

esca <- esca %>% select(seq_num, diagnosis, dilute_library_concentration)

hbv <- hbv %>% select(seq_num, diagnosis, dilute_library_concentration)

hcc <- hcc %>% select(seq_num, diagnosis, dilute_library_concentration)

stad <- stad %>% select(seq_num, diagnosis, dilute_library_concentration)

hea <- hea %>% select(seq_num, diagnosis, dilute_library_concentration)

cohort <- rbind(brca, crc, gbm, esca, hbv, hcc, stad, hea)

# There are blank entries in the cohort - but merging by seq_num and diagnosis should take care of past problems with having multiple seq_num ids
#View(cohort)

# Add SEQ as a prefix
cohort$seq_num <- paste0('SEQ', cohort$seq_num)

# In the new pheno dataset - some of the healthy patients were given the label CRN to identify non-colon cancers. These are also healthy patients and so I am going to change the label as such
mc_data$Cat[mc_data$Cat == "CRN"] <- "HEA"

# Merge with mc_data
colnames(mc_data)[2] <- "diagnosis"
mc_data <- inner_join(mc_data, cohort, by = c("seq_num", "diagnosis"))
```

Training and Testing Splits for all Cancers Separate

Create balanced datasets:
https://topepo.github.io/caret/data-splitting.html
```{r}
set.seed(02051)
trainIndex <- createDataPartition(mc_data$diagnosis, p = .7, 
                                  list = FALSE, 
                                  times = 1)
head(trainIndex)

mcTrain <- mc_data[trainIndex,]
mcTest  <- mc_data[-trainIndex,]

table(mc_data$diagnosis)
table(mcTrain$diagnosis)
table(mcTest$diagnosis)

round(table(mc_data$diagnosis)/342*100, 1)
round(table(mcTrain$diagnosis)/242*100, 1)
round(table(mcTest$diagnosis)/100*100, 1)
# Yes, looking at percents above shows equal ditributions
```

Check if any missing values: none!
```{r}
sum(is.na(mcTrain))
sum(is.na(mcTest))
```

Combined Colon and Stomach Cancer 

```{r}
# Copy training and testing datasets into new dataframes
mcTrain_cs <- mcTrain
mcTest_cs <- mcTest

# combine Colon and Stomach Cancer
mcTrain_cs$diagnosis[mcTrain_cs$diagnosis == "CRC" | mcTrain_cs$diagnosis == "STAD"] <- "CRC_STAD"

mcTest_cs$diagnosis[mcTest_cs$diagnosis == "CRC" | mcTest_cs$diagnosis == "STAD"] <- "CRC_STAD"

table(mcTrain_cs$diagnosis)
table(mcTest_cs$diagnosis)

round(table(mcTrain_cs$diagnosis)/242*100, 1)
round(table(mcTest_cs$diagnosis)/100*100, 1)
# Yes, looking at percents above shows equal ditributions
```

Check if any missing values: none!
```{r}
sum(is.na(mcTrain_cs))
sum(is.na(mcTest_cs))
```

Save training and test sets
```{r}
write.csv(mc_data, file = "C:/Users/User/Box Sync/Projects/Multi_Cancer_DL/02_Processed_Data/mc_data.csv", row.names = FALSE)

write.csv(mcTrain, file = "C:/Users/User/Box Sync/Projects/Multi_Cancer_DL/02_Processed_Data/mcTrain_70_30.csv", row.names = FALSE)

write.csv(mcTest, file = "C:/Users/User/Box Sync/Projects/Multi_Cancer_DL/02_Processed_Data/mcTest_70_30.csv", row.names = FALSE)

write.csv(mcTrain_cs, file = "C:/Users/User/Box Sync/Projects/Multi_Cancer_DL/02_Processed_Data/mcTrain_cs_70_30.csv", row.names = FALSE)

write.csv(mcTest_cs, file = "C:/Users/User/Box Sync/Projects/Multi_Cancer_DL/02_Processed_Data/mcTest_cs_70_30.csv", row.names = FALSE)
```
